File: mixed/num_methods/2course2sem/rungeKutta/adams4.m
function Y = adams4(X, h, f, y0)
%X is vector of watch points, X = a:h0:b, h >= h0, h0/h in N (last - adviceable)
%e. g. suppose X = [-1 0 1], h = 0.5
%then you calculate solution Y1 on net X1 = [-1 -0.5 0 0.5 1]
%and output values Y only corresponding X points
X1 = X(1):h:X(end); %dense net
Y1 = zeros(size(X1));
Y1(1) = y0;
for i = 1:1:3 %boosting with RungeKutta
    n1i = f(X1(i), Y1(i));
    n2i = f(X1(i) + h/2, Y1(i) + (h/2)*n1i);
    n3i = f(X1(i) + h/2, Y1(i) + (h/2)*n2i);
    n4i = f(X1(i) + h, Y1(i) + h*n3i);
    dyi = (h/6)*(n1i+2*n2i+2*n3i+n4i);
    Y1(i+1) = Y1(i) + dyi; 
    %Y1(i+1) = Y1(i) + h*f(X1(i), Y1(i)); %simple Euler's method
end

for i = 4:1:size(X1, 2)-1
    f1 = f(X1(i), Y1(i));
    f2 = f(X1(i-1), Y1(i-1));
    f3 = f(X1(i-2), Y1(i-2));
    f4 = f(X1(i-3), Y1(i-3));
    dyi = (h/24)*(55*f1-59*f2+37*f3-9*f4);
    Y1(i+1) = Y1(i) + dyi; 
    %Y1(i+1) = Y1(i) + h*f(X1(i), Y1(i)); %simple Euler's method
end

sX = size(X, 2);
sX1 = size(X1, 2);
delta = round(sX1 / sX);
Y = zeros(size(X));

for i = 1:1:sX
    if delta*i <= sX1
        Y(i) = Y1(delta*(i-1)+1);
    else
        disp('out of bounds')
    end
    disp(delta)
end
end

File: mixed/num_methods/2course2sem/rungeKutta/adams4.cpp
#include"rungeKutta.h"

vector <double> Solution::adams4(vector<double> X, int n) {
	double h = (b0 - a0) / (n - 1);
	vector<double> X1;
	for (int i = 0; i < n; i++)
		X1.push_back(a0 + h * i);
	vector<double> Y1;
	Y1.resize(n);
	Y1[0] = y0;
	double n1i, n2i, n3i, n4i, dyi;
	for (int i = 0; i < 3; i++) {
		n1i = f(X1[i], Y1[i]);
		n2i = f(X1[i] + h / 2, Y1[i] + (h / 2)*n1i);
		n3i = f(X1[i] + h / 2, Y1[i] + (h / 2)*n2i);
		n4i = f(X1[i] + h, Y1[i] + h * n3i);
		dyi = (h / 6)*(n1i + 2 * n2i + 2 * n3i + n4i);
		Y1[i + 1] = Y1[i] + dyi;
	}
	double f1, f2, f3, f4;
	for (int i = 3; i < n - 1; i++) {
		f1 = f(X1[i], Y1[i]);
		f2 = f(X1[i - 1], Y1[i - 1]);
		f3 = f(X1[i - 2], Y1[i - 2]);
		f4 = f(X1[i - 3], Y1[i - 3]);
		dyi = (h / 24)*(55 * f1 - 59 * f2 + 37 * f3 - 9 * f4);
		Y1[i + 1] = Y1[i] + dyi;
	}
	int delta = (int)(n - 1) / (X.size() - 1);
	vector <double> Y;
	for (int i = 0; i < X.size(); i++) {
		if (delta * i < n) Y.push_back(Y1[delta*i]);
	}
	return Y;
}

